### PR TITLE
protocol/serial: add serial protocol implementation

### PR DESCRIPTION
**Pull request by csssuf**
Also pulls in bitflags as a dependency in order to represent the ControlBits structure.

### CODE REVIEW

File: src/protocol/serial.rs [hunk 1]
```diff
@@ -0,0 +1,215 @@
...
+    /// Write a string to the serial device.
+    pub fn write(&self, data: &str) -> Result<usize, Status> {
+        self.write_raw(data.as_ptr(), data.len())
+    }
+
+    fn read_raw(&self, buf: *mut u8, length: usize) -> Result<Option<(*const u8, usize)>, Status> {
+        let mut new_length = length.clone();
+        unsafe {
+            match (self.read)(self, &mut new_length, buf as *mut CVoid) {
+                Status::Success => Ok(Some((buf, new_length))),
```
Comment (bgilbert) [inline src/protocol/serial.rs]: I assume you're returning `buf` for chaining purposes?  If so, is that a sufficient reason to do it?

File: src/protocol/serial.rs [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+            .allocate_pool(length)
+            .and_then(|buf_ptr| self.read_raw(buf_ptr, length))
+    }
+
+    /// Read a `length`-byte string from the serial device. The bytes will be interpreted as all
+    /// UTF-8 characters, since the serial device should only be returning 7-bit ASCII data.
+    pub fn read_str(&self, length: usize) -> Result<Option<&str>, Status> {
+        self.read_bytes(length).map(|option| {
+            option.map(|(bytes, read_length)| unsafe {
+                str::from_utf8_unchecked(slice::from_raw_parts(bytes, read_length))
```
Comment (bgilbert) [inline src/protocol/serial.rs]: Is this an in-place conversion?  If not, you're leaking the buffer.

The user's charset is not necessarily UTF-8 and there's nothing preventing them from emitting the high bit, so `unchecked` is not safe.

**Review by bgilbert — CHANGES_REQUESTED** [review]

File: src/protocol/serial.rs (hunk 1 above)
Comment (csssuf) [reply src/protocol/serial.rs]: It is for chaining, yes. I think it's slightly cleaner to do so but I'm not overly attached to it, is there a reason not to?

File: src/protocol/serial.rs (hunk 2 above)
Comment (csssuf) [reply src/protocol/serial.rs]: The conversion is in-place, so there's no leak. I think the best approach here might be to just expose a `read_bytes` which returns a slice of bytes instead, and let library consumers do with the bytes what they want.

File: src/protocol/serial.rs [hunk 3]
```diff
@@ -0,0 +1,261 @@
...
+impl SerialIOProtocol {
+    pub fn new(timeout: u32) -> Result<SerialIOProtocol, Status> {
+        let bs = ::get_system_table().boot_services();
+        bs.locate_protocol::<RawSerialIOProtocol>(0 as *const CVoid)
+            .and_then(|protocol| {
+                if let Err(e) = protocol.reset() {
+                    return Err(e);
+                }
+
+                let attr = protocol.set_attributes(None, None, Some(timeout), None, None, None);
```
Comment (bgilbert) [inline src/protocol/serial.rs]: Presumably we need to support non-default configuration?

File: src/protocol/serial.rs [hunk 4]
```diff
@@ -0,0 +1,261 @@
...
+                }
+                Ok(SerialIOProtocol {
+                    raw_protocol: protocol,
+                    timeout: timeout
+                })
+            })
+    }
+
+    pub fn write(&self, data: &str) -> Result<usize, Status> {
+        let attr_status = self.raw_protocol.set_attributes(None, None, Some(self.timeout), None, None, None);
```
Comment (bgilbert) [inline src/protocol/serial.rs]: Needs a comment explaining why we do this before every I/O.

**Review by bgilbert — CHANGES_REQUESTED** [review]

File: src/protocol/serial.rs [hunk 5]
```diff
@@ -0,0 +1,307 @@
...
+        self.parity = parity.or(self.parity);
+        self.data_bits = data_bits.or(self.data_bits);
+        self.stop_bits = stop_bits.or(self.stop_bits);
+
+        self.set_attributes()
+    }
+
+    /// Write a string to the serial device.
+    pub fn write(&self, data: &str) -> Result<usize, Status> {
+        // GRUB sets the attributes on the serial device on each read or write, so we will too.
```
Comment (bgilbert) [inline src/protocol/serial.rs:293]: We're also doing it because it avoids a specific problem.  Please document what that is.

**Review by bgilbert — APPROVED** [review]

**Pull request merged**